### ISSUE TITLE
Repo: Write lint rule to prefer tsutils methods over bitwise checks

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#11625]
### Suggestion

Coming over from #11617: it's not well-known that [`ts-api-utils`](https://github.com/JoshuaKGoldberg/ts-api-utils) provides a handful of shortcut/wrapper methods around TypeScript bitwise enums/flags. For example, this method:

```ts
function hasBaseTypes(type: ts.Type): type is ts.InterfaceType {
  return (
    (type.flags & ts.TypeFlags.Object) !== 0 &&
    (((type as ts.ObjectType).objectFlags & ts.ObjectFlags.Interface) !==
      0 ||
      ((type as ts.ObjectType).objectFlags & ts.ObjectFlags.Class) !== 0)
  );
}
```

...can be rewritten as:

```ts
function hasBaseTypes(type: ts.Type): type is ts.InterfaceType {
  return (
    tsutils.isObjectType(type) &&
    tsutils.isObjectFlagSet(
      type,
      ts.ObjectFlags.Interface | ts.ObjectFlags.Class,
    )
  );
}
```

Proposal: let's write a general-purpose lint rule that enforces the latter when possible?

### Additional Info

Eventually this might make sense to move to https://github.com/eslint-community/eslint-plugin-eslint-plugin, perhaps.

💖

**Comment by kirkwaiblinger** [comment]
> when possible

Detection might be tricky? 

Seems reasonable to lint for if feasible, though.

**Comment by Kimsoo0119** [comment]
Hi! I'd like to work on this issue.

Before I start, I want to clarify the scope: 

1. Should this rule be implemented in `eslint-plugin-internal` for now?
2. Which patterns should the rule detect? For example:
    - `type.flags & ts.TypeFlags.X` → `isTypeFlagSet(type, ts.TypeFlags.X)`
    - `(type as ts.ObjectType).objectFlags & ts.ObjectFlags.X` → `isObjectFlagSet(type, ...)`
    - Are there other patterns to consider (e.g., `SymbolFlags`)?
3. Should the rule provide auto-fix suggestions, or just report errors?
            
Thanks!

**Comment by JoshuaKGoldberg** [comment]
> 1. Should this rule be implemented in `eslint-plugin-internal` for now?

Yes.

> 2. Which patterns should the rule detect? For example: ...

I think all three of those are a great start. If you search for lines that have `.flag`, `&`, and a numeric literal such as `0`, those three are most-to-all of them.

> 3. Should the rule provide auto-fix suggestions, or just report errors?
> 
> Thanks!

I think auto-fix 👍 that'll make it easier to fix things internally. We expect PRs to pass linting and also enable any new lint rules, so this'll save you some time I hope!
